test/suite/test_prepare22.py

```diff
@@ -0,0 +1,137 @@
+#!/usr/bin/env python
+#
+# Public Domain 2014-present MongoDB, Inc.
+# Public Domain 2008-2014 WiredTiger, Inc.
+#
+# This is free and unencumbered software released into the public domain.
+#
+# Anyone is free to copy, modify, publish, use, compile, sell, or
+# distribute this software, either in source code form or as a compiled
+# binary, for any purpose, commercial or non-commercial, and by any
+# means.
+#
+# In jurisdictions that recognize copyright laws, the author or authors
+# of this software dedicate any and all copyright interest in the
+# software to the public domain. We make this dedication for the benefit
+# of the public at large and to the detriment of our heirs and
+# successors. We intend this dedication to be an overt act of
+# relinquishment in perpetuity of all present and future rights to this
+# software under copyright law.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+# EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+# MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
+# IN NO EVENT SHALL THE AUTHORS BE LIABLE FOR ANY CLAIM, DAMAGES OR
+# OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
+# ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
+# OTHER DEALINGS IN THE SOFTWARE.
+
+import wiredtiger, wttest
+from wtscenario import make_scenarios
+
+# test_prepare22.py
+# Test rollback to stable clears the WT_UPDATE_TO_DELETE_HS flag.
```

Comment:
I'm concerned with how specific this test case is. It feels like we don't have enough stress test coverage across this type of functionality.
Traditionally we've relied on test/format giving us coverage for this type of (corner case) scenario. Can you review test/format and give an opinion about what might be missing from it that means we didn't see this failure there?

Reply:
The test I wrote is for the [second scenario](https://jira.mongodb.org/browse/WT-9592?focusedCommentId=4694022&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-4694022) I wrote in the ticket, which I confirm is not happening. I include the test anyway in the pr to ensure it will not fail in the future.

The real issue is [scenario 1](https://jira.mongodb.org/browse/WT-9592?focusedCommentId=4691000&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-4691000).

Here's I think why test format doesn't catch scenario 1. It is a very specific case:

1. We need cache pressure.
3. we need to have a failed eviction after we have successfully built the disk image with the prepared update.
4. We need to rollback the prepared update.
5. We need to have a failed reconciliation after we have successfully built the disk image so the whole update chain is restored back.
6. We need to run rollback to stable at this moment before the page is reconciled again.

It is currently hard to write a reliable python test for this case because we cannot control eviction to fail after successfully built the disk image.
 
In test format, we may experience cache pressure but I don't think we run rts on the live tree very often.

Reply:
@quchenhao that is exactly the type of scenario test/format historically found for us.

Could you check whether it does RTS at runtime? If not, whether it would be possible to add it in?

Reply:
```
    /*
     * The system should be quiescent at this point, call rollback to stable. Generally, we expect
     * applications to do rollback-to-stable as part of the database open, but calling it outside of
     * the open path is expected in the case of applications that are "restarting" but skipping the
     * close/re-open pair. Note we are not advancing the oldest timestamp, otherwise we wouldn't be
     * able to replay operations from after rollback-to-stable completes.
     */
    rollback_to_stable();

    if (lastrun) {
        tinfo_teardown();
        timestamp_teardown(session);
    }
```
Test format is running rollback to stable but only after we have done all the normal operations. To trigger the issue, we also need WiredTiger to do another reconciliation after rollback to stable. May be we can force to add a checkpoint after the rollback_to_stable() call.

Reply:
Correction, we do rts for each set of operations. It may run multiple times in a single run.

```
    /*
     * Calculate how long each operations loop should run. Take any timer value and convert it to
     * seconds, then allocate 15 seconds to do initialization, verification and/or salvage tasks
     * after the operations loop finishes. This is not intended to be exact in any way, just enough
     * to get us into an acceptable range of run times. The reason for this is because we want to
     * consume the legitimate run-time, but we also need to do the end-of-run checking in all cases,
     * even if we run out of time, otherwise it won't get done. So, in summary pick a reasonable
     * time and then don't check for timer expiration once the main operations loop completes.
     */
    ops_seconds = GV(RUNS_TIMER) == 0 ? 0 : ((GV(RUNS_TIMER) * 60) - 15) / FORMAT_OPERATION_REPS;
    for (reps = 1; reps <= FORMAT_OPERATION_REPS; ++reps)
        operations(ops_seconds, reps == FORMAT_OPERATION_REPS);
```

Reply:
It seems like test/format should have all the necessary ingredients then. Could you try to construct a case where format will fail?
Maybe it needs the mirror branch, to be able to detect the problem?